run examples with all enabled accelerators

psychocoderHPC opened the pull request:
Remove the usage of `ExampleDefaultAcc` in all examples.
 Examples will now be executed for all enabled accelerators. 
Fix that `ExampleDefaultAcc` was used in the test `MapIdxPitchBytes`.

note: The stream benchmark is still using `ExampleDefaultAcc`, this will be addressed in a separate PR.

psychocoderHPC committed:
fix windows compile error

Add workaround for not supported type deduction in the `noexcept()`
expression.

mehmetyusufoglu commented on example/bufferCopy/src/bufferCopy.cpp:
```diff
@@ -257,3 +257,9 @@ auto main() -> int
+auto main() -> int
+{
+    // execute the example once for each enabled accelerator
+    return std::apply([](auto const&... tags) { return (example(tags) || ...); }, alpaka::EnabledAccTags{});
```
Users will be challenged and might think that this is the way alpaka programs are written. It is an extra layer of difficulty for a novice user. ( I thought a solution by using `std::variant `and `std::visit `but seems not simpler than this solution. )

mehmetyusufoglu replied:
Can't we solve this at CI level?

mehmetyusufoglu replied:
Call `ctest -v`  many times in run_test.sh with different arguments? ( Add arguments to example's main function (i.e. argv) in cpp and use  CMAKE_CTEST_ARGUMENTS somehow? )

psychocoderHPC replied:
Adding arguments will increase the complexity much more. The problem is we write code which can be executed on many devices so it is not like other projects where a simple main is enough.
This PR solves an issue that we are not able to execute our examples on all devices.
The now-shown CI issues show how easy it is to run the example with all accelerators.

fwyzard replied:
> Users will be challenged and might think that this is the way alpaka programs are written. It is an extra layer of difficulty for a novice user.

I fully agree.

If you want to write tests like this, OK... I personally strongly dislike this, but I won't object to obscure unreadable code in the tests, as long as the compiler and CI are happy.

But this should definitely not be how we write an example.

SimeonEhrig commented on example/bufferCopy/src/bufferCopy.cpp:
```diff
@@ -64,7 +63,8 @@ struct FillBufferKernel
 };
 
-auto main() -> int
+template<typename TAccTag>
```
Maybe you should a small comment.

```c++
// In standard projects, not every kernel is executed with all active accelerators.
// Instead, a single accelerator is selected once from the active accelerators and the kernels are executed on the selected accelerator.
// If you use the example as the starting point for your project, you can rename the example() function to main() and move the accelerator type to the function body.
```

fwyzard commented:
@psychocoderHPC, please do not make the _examples_ even harder to understand :-(

psychocoderHPC commented:
> @psychocoderHPC, please do not make the _examples_ even harder to understand :-(

But I like to solve the issue we have that only a magic accelerator is used and as you can see the examples was not able to run with other accelerators because we never tested it.

fwyzard commented:
> But I like to solve the issue we have that only a magic accelerator is used and as you can see the examples was not able to run with other accelerators because we never tested it.

I don't disagree.

Do you think we could move the logic to build and test different accelerators to the build system ?

psychocoderHPC commented:
> Do you think we could move the logic to build and test different accelerators to the build system ?

I thought about it to but it look to me very complicated and will add a lot of code.

I decided to restructure my changes and add a helper to hide the folding expression.

psychocoderHPC committed:
Remove the usage of `ExampleDefaultAcc` in all examples.

Examples will now be executed for all enabled accelerators.
Fix that `ExampleDefaultAcc` was used in the test `MapIdxPitchBytes`.

psychocoderHPC commented:
I updated all examples and hopefully simplified the main by removing the folding expression.

SimeonEhrig commented on example/bufferCopy/src/bufferCopy.cpp:
```diff
@@ -257,3 +246,19 @@ auto main() -> int
+    //   TagCpuSerial, TagGpuHipRt, TagGpuCudaRt, TagCpuOmp2Blocks, TagCpuTbbBlocks,
+    //   TagCpuOmp2Threads, TagCpuSycl, TagCpuTbbBlocks, TagCpuThreads,
+    //   TagFpgaSyclIntel, TagGenericSycl, TagGpuSyclIntel
+    return alpaka::executeForEachAccTag([=](auto const& tag) { return example(tag); });
```
To honest, I'm not sure if we should use tags directly. Now it looks like, that tags are required to develop an alpaka application.

Maybe we should use the tags only internal in the `alpaka::executeForEachAccTag` function, which should become `alpaka::executeForEachAcc`.

psychocoderHPC replied:
I do not see a way to implement it as easy as it is now. The reason is that we can not create an instance of Acc, so we can not pass it to the execution lambda.
I also prefer that we advertise tags more because of the reason that you are not allowed to instantiate tags on the host side. This avoids using accelerators as an argument in a lambda.
Additionally accelerators have the problem that the definition are not available if the backend is not activated in CMake.

psychocoderHPC replied:
There is another advantage of using tags. Tags only say to which accelerator you would like to map your problem but is not say anything about the dimensionality of the index domain. This provides the freedom to switch dimensionality for each kernel depending on the needs.

SimeonEhrig replied:
```suggestion
    return alpaka::executeForEachAccTag([=](auto const& tag) { return example<alpaka::TagToAcc<decltype(TagToAcc)>>(); });
```

```c++
template<typename TAcc>
auto example() -> int {
// ...
using Acc = TAcc;
// ..
}
```

psychocoderHPC replied:
IMO

```C++
return example(tag);
```
vs.
```C++
return example<alpaka::TagToAcc<decltype(TagToAcc)>>();
```

The problem is any way that the dimension and index type for the accelerator is required to create an accelerator.
So it will be 

```C++
return example<alpaka::TagToAcc<decltype(TagToAcc), Dim, Idx>();
```

Where Dim and Idx must moved away from the problem into the main function.

psychocoderHPC commented:
@alpaka-group/alpaka-maintainers please merge

fwyzard commented on include/alpaka/example/ExecuteForEachAccTag.hpp:
```diff
@@ -0,0 +1,27 @@
+    {
+        // Execute the callable once for each enabled accelerator.
+        // Pass the tag as first argument to the callable.
+        return std::apply([=](auto const&... tags) { return (callable(tags) || ...); }, alpaka::EnabledAccTags{});
```
@psychocoderHPC sorry, I hadn't thought about this before: does the `||` operator obey the usual short-circuit rules here ?

That is, as soon as one of the `callable` returns `true`, will the rest be skipped ?

psychocoderHPC replied:
Yes short-circuit should be active but keep in mind that return zero is the case where we like to execute all. In case non-zero is returned what is an error we will stop the execution.

fwyzard replied:
OK, so we don't want to continue running with the other tags if one fail ?

psychocoderHPC replied:
Yes exact, I think we should not waiste compute time if we failed. 
I failure can also bring the device into a state that all followup work fails, in that case the last error will only be an artefact of the failing first run but the user will first look to the last failure.
